### DISCUSSION TITLE
Question about the two ways to define methods

### DISCUSSION THREAD

**Discussion by ocavue** [remirror/remirror#360] [Ideas] [answered]
Hi @ifiokjr. I'm migrating to `remirror@next` for `Rino`. I found that there are two ways to define a method in an `Extension`: as a *property* or as a *function*.

```js
class MyExtension extends NodeExtension {
    createNodeSpec(extra) { ... } // function
    createKeymap = () => { ... } // property
}
```

It seems that all methods defined by extra extensions will use the *property* way (`KeymapExtension` for `createKeymap`, `CommandsExtension` for `createCommands`), but I don't fully understand the reason behind this design. Would you give me some explanation or hint about it?

**Comment by ifiokjr** [comment]
Good question. Honestly the main reason for this design is this [method signature eslint rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/method-signature-style.md). 

It enforces property signatures for maximum strictness. However, it has unintended effects in recent versions of TypeScript. The compiler will display an error if using a property function declaration when the type signature is a property signature.

### property signature for functions

- **Pros**
  - Stricter types
  - All methods have `this` scoped to the class instance.

- **Cons**
  - Property methods are not defined on the prototype and are not shared across instances so there is a slight performance penalty
  - Property methods that need to be used in the constructor must be called after `super` when extending another class
  - Forces users to use property methods which may not be inline with the style of their codebase.

### method signature for functions

- **Pros**
  - End users can use both properties and methods
  - Allows methods to be defined on the class prototype for slight performance gains

- Cons
  - Less type strictness. 



---

Listing out the pro's and cons makes me want to change to using methods instead of properties. It's a pretty big change, but I'd be happy to see it happen.

**Comment by ocavue** [comment]
Another con for property signature: it does not work with inheritance (we can not use super in an arrow function to call the same function of a super class).

```ts
class A {
    f = () => {
        return "a"
    }

    g() {
        return "aa"
    }
}

class B extends A {
    f = () => {
        const s = super.f() // will raise error
        return "b" + s
    }

    g() {
        const s = super.g()
        return "bb" + s
    }
}

    const a = new A()
    const b = new B()
    console.log({
        'a.f': a.f(),
        'a.g': a.g(),
        //'b.f': b.f(), // error
        'b.g': b.g(),
    })

```

**Reply by ifiokjr** [reply]
@ocavue  How comfortable would you be in refactoring the codebase to use methods instead of properties?

**Reply by ocavue** [reply]
I am happy to do this job.

I should be able to submit a PR for the first part of the refactoring in this weekend.

**Reply by ifiokjr** [reply]
Awesome!

I was a bit naughty in that I had started doing some bits earlier on, but I'll push up what I've done so far as it might be a good place to start from. I'll update this when I've pushed the update.

**Comment by ifiokjr** [comment, marked as answer]
#377 has been assigned to you 😊 

I haven't yet updated the extension lifecycle methods in `packages/@remirror/core/src/extension/extension-base.ts`. 

They should no longer pass in the first parameter `extensions: AnyExtension[]`. I think this is better since the extension list is accessible via `this.store.extensions`.

```ts
interface ExtensionLifecycleMethods {
  /**
   * This handler is called when the `RemirrorManager` is first created.
   *
   * @remarks
   *
   * Since it is called as soon as the manager is some methods may not be
   * available in the extension store. When accessing methods on `this.store` be
   * sure to check when they become available in the lifecycle. It is
   * recommended that you don't use this method unless absolutely required.
   */
  onCreate?(): void;

  /**
   * This event happens when the view is first received from the view layer
   * (e.g. React).
   */
  onView?(view: EditorView<EditorSchema>): void;

  /**
   * Called when a transaction successfully updates the editor state.
   */
  onStateUpdate?(parameter: StateUpdateLifecycleParameter): void;

  /**
   * Called when the extension is being destroyed.
   */
  onDestroy?(): void;
}
```

I'd also like to remove the lifecycle types and update all affected code.

```diff
- export type CreateLifecycleMethod = (extensions: readonly AnyExtension[]) => void;
- export type ViewLifecycleMethod = (
-   extensions: readonly AnyExtension[],
-   view: EditorView<EditorSchema>,
- ) => void;
- export type DestroyLifecycleMethod = (extensions: readonly AnyExtension[]) => void;
```
